sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala

```diff
@@ -359,10 +359,27 @@ case class LambdaVariable(value: String, isNull: String, dataType: DataType) ext
 case class MapObjects(
     function: AttributeReference => Expression,
     inputData: Expression,
-    elementType: DataType) extends Expression {
+    elementType: DataType)(
+    completeFunctionCopy: Option[Expression] = None) extends Expression {
```

Comment:
How about we do this instead?  Make the `MapObjects` class have both `loopVar` and `function` as constructor arguments.  A custom `apply` method will take `AttributeReference => Expression` and construct the object by calculating the complete function eagerly.